acvm-repo/acvm/src/pwg/mod.rs

```diff
@@ -737,9 +752,30 @@ fn any_witness_from_expression<F>(expr: &Expression<F>) -> Option<Witness> {
 pub(crate) fn is_predicate_false<F: AcirField>(
     witness: &WitnessMap<F>,
     predicate: &Option<Expression<F>>,
+    pedantic_solving: bool,
+    opcode_location: &ErrorLocation,
 ) -> Result<bool, OpcodeResolutionError<F>> {
     match predicate {
-        Some(pred) => get_value(pred, witness).map(|pred_value| pred_value.is_zero()),
+        Some(pred) => {
+            let pred_value = get_value(pred, witness);
+            if pedantic_solving {
+                pred_value.and_then(|predicate_value| {
+                    if predicate_value.is_zero() {
+                        Ok(true)
+                    } else if predicate_value.is_one() {
+                        Ok(false)
+                    } else {
+                        let opcode_location = *opcode_location;
+                        Err(OpcodeResolutionError::PredicateLargerThanOne {
+                            opcode_location,
+                            predicate_value,
+                        })
+                    }
+                })
+            } else {
+                pred_value.map(|pred_value| pred_value.is_zero())
+            }
+        }
```

Comment:
```suggestion
            let pred_value = get_value(pred, witness)?;
            let predicate_is_false = pred_value.is_zero();
            if pedantic_solving {
               // We expect that the predicate should resolve to either 0 or 1.
                if !predicate_is_false && !predicate.is_one() {
                     let opcode_location = *opcode_location;
                        return Err(OpcodeResolutionError::PredicateLargerThanOne {
                            opcode_location,
                            predicate_value,
                        })
                }
            }
            Ok(predicate_is_false)
```

I think we can separate the calculation of the result from the pedantic check.